Rewrite it in rust (regex, utils) (#220)
y-scope/log-surgeon: rust/docs/parsing.md

## Parsing in Log Surgeon
An overview of the title.

> See also: [Parsing Specification File][parsing-spec].

Reviewer: ```suggestion
Log Surgeon parses log data using a [Parsing Specification File][parsing-spec]. This includes:
- How root rules are matched.
- How subrule matches are extracted.
- How log events are separated.

Parsing consists of two main stages:

1. A parser that advances through the input and builds structured log events. 
2. A lexer that identifies matches in the input according to the parsing specification.
```

> 

Reviewer: Add an index.

Author: IMO the whole parsing doc probably ought to/will be rewritten soon with the various things we want to change with parsing, like I can add links to "matching root rules", "implementation details", "submatch extraction", and "future work", but it kinda seems pointless?

Reviewer: I think its worth just having organizationally for 2 reasons:
- If we have to focus on other things first, and this version ends up being slightly longer term than intended it is going to be pushed to the main repo, so it should be in some finalized state.
- The index being there acts as a hint to the next reviewer that the index needs to be updated when the changes are made.

Author: Ok - I removed the future work section though, that was reference for the paper and already out of date

> ### Matching Root Rules

Reviewer: ```suggestion
### Matching Root Rules
The lexer identifies occurrences of root rules defined in the parsing specification. A root rule match represents a piece of semantically meaningful text identified by the user.

```

> A lexer processes input left to right and reports root rule matches.

Reviewer: This refers to a root rule, but to know what that is the user has to refer to the spec readme, and even inside that its not easy to find, I think it might be useful to have a glossary, especially since terminology for this stuff is pretty niche / honestly just decided by us, so we can't expect the user to know any of it.

At each step, if possible, the lexer takes the longest possible match of any root rule.

> If multiple root rules match with the same length, the highest-priority (earliest) rule is returned.

Reviewer: ```suggestion
The lexer processes input left to right. At each step, it attempts to match root rules according to the following:
- The longest possible match.
- If multiple root rules match with the same length, the highest-priority (earliest) rule in the spec is selected.
```

If no root rules match at the current position,

> the parser seeks to the first delimiter character after the current position,

Reviewer: Should make a git issue about turning this into seeking out the first matchable character (whether that character is based on an anchor or different regex), and then we can relax the delimiter requirement.

> and repeats attempting to match a root rule _after_ the delimiter.

Reviewer: ```suggestion
and the lexer repeats attempting to match a root rule _after_ the delimiter.
```

WIP: We hope to generalize root rule matching to find the earliest possible occurrence of a root rule at each step.

#### Implementation Details
To determine if/what rule matches from an input position,
Log Surgeon builds an [automaton][dfa] for the combination of all root rule patterns in the parsing specification;
an automaton is just a state machine with transitions based on an input character.
Specifically, Log Surgeon implements the classical regex -> NFA -> DFA construction.

DFAs simulate matching multiple patterns/rules at once with just a single pass through the input.

> "Executing" the DFA is just a loop that traverses the states:

Reviewer: ```suggestion
DFAs simulate matching multiple patterns/rules at once performing only a single pass through the input.
"Executing" the DFA is a loop that traverses the states:
```

```rust
let mut current_state: usize = 0;
let mut maybe_match: Option<(Rule, usize)> = None;
for (pos, ch) in input.char_indices() {
	if let Some(next_state) = states[current_state].lookup_transition(ch) {
		current_state = next_state;
		if let Some(rule) = states[current_state].accepting_rule {
			// Save the match, but don't stop yet; see if we can match longer.
			maybe_match = Some((rule, pos + ch.len_utf8()));
		}
	} else {
		break;
	}
}
return maybe_match;
```

> Note: The theoretical complexity of determining whether input matches
> is very different from finding the longest possible match;
> longest match semantics inherently requires "looking ahead" to attempt a longer match,

Reviewer: ```suggestion
Note: The theoretical complexity of determining the longest possible match requires "looking ahead",
```

Author: I kinda want to emphasize the two different operations, since IMO it's easy to conflate the two and think there's a mistake

even if the extra input does not result in a match.
In practice, the non-static text of logs is small compared to the static text,
and confirming the longest possible match for rules in a parsing specification
rarely require consuming much extra input.

Notice that each iteration of the loop requires looking up the transition for the current state and input character.

> In fact, transitions for a state are stored as a list of (non-overlapping) Unicode code point intervals,

Reviewer: Is there no optimization for ascii only input? Or does the JIT kinda make that redundant?

Author: For the "plain rust implementation" there is optimization for ascii characters (in `lookup_transitions`); I didn't mention it here since it's not really fundamental. For the JIT, adding a hot path for ascii actually made it slower (idk, hardware reasons), so the JIT doesn't do it

Reviewer: Is it not important to mention here since you're pointing out the cost of using Unicode? Maybe something in passing like, this is slightly mitigated in a lot of cases which as ascii only.

Author: I rewrote it slightly to be simpler/omit the part about intervals; the doc had more implementation details for reference for the paper, but I don't know if this part needs to spell out the code so much. I think its worth keeping in Unicode code points since that's more meaningful w.r.t. the fact that it is encoding aware (even if implementation only currently supports utf-8), and I added the ASCII note in parentheses

so this lookup means comparing the input character against these intervals.
Further, at each step, we need to check if the current state is an accepting state.
In a sense, this DFA loop is an "interpreter" for instructions "goto next state" and "record match".

To speed up matching, we compile DFAs into native functions.
Currently, Log Surgeon just in time (JIT) compiles them for ease of deployment,

> but conceptually the DFAs are the same as if they were "ahead of time" compiled.

Reviewer: If its easy, maybe add 1 sentence summarizing the fact that JIT removes transitions lookup and state management by encoding the DFA into the native flow control, or w.e. the more technically correct phrasing is. Although, it is explained in detail below.

we simply don't emit `record_match()` for a state if it isn't an accepting state.

##### Submatch Extraction
The classical DFA execution determines which rule matches;
once the rule and matched text is known,
Log Surgeon uses a [tagged DFA][tagged-dfa] for the specific rule to extract sub-rule matches.
Conceptually, a tagged DFA is just a DFA with operations to execute on state transitions;
in this case, the operations record sub-rule match positions.
Compared to the pseudocode for a classical DFA above,
the inner loop just changes by:

```rust
if let Some((next_state, operations)) = states[current_state].lookup_transition(ch) {
	execute_operations(operations);
	// Rest is the same.
	current_state = next_state;
	if let Some(rule) = states[current_state].accepting_rule {
		maybe_match = Some((rule, pos + ch.len_utf8()));
	}
} else {
	break;
}
```

Note: While it is possible to build a single tagged DFA for all the rules in the parsing specification,
executing this DFA to determine the root and sub-rule matches at once means
executing all the state transition operations for all the potential rule matches;

> i.e. recording many potential sub-rule matches that are discarded.

Reviewer: This seems weird, maybe open a git issue to explore this farther. I feel like we should get concrete numbers on the single tagged DFA approach and two pass approach and explore a bit. It may also be something that benefits a lot from a larger lookahead to try to prevent tag operations.

Author: FWIW the single tagged DFA was the original approach and was much much slower - of course, there's room for optimization, but the problem was with a single tagged DFA, it had to execute tags for every "still matchable" rule, and so matching got slower with larger schemas (theoretically, if there were no overlaps, it shouldn't matter)

I can make an issue though IMO this is lower prio to try compared to other things we want to change with parsing (e.g. grep semantics, parsing based on log types)

Reviewer: I think its worth keeping track. I was originally surprised we weren't hitting slow downs with more tagging like in the c++ version, but now I understand its cause we do a 2 pass approach. I think its just worth keeping the idea open of exploring the 1 pass approach.

Therefore, we have found it better to execute a classical DFA to determine which rule,
and a tagged DFA specifically for the matched rule.

#### Future Work

> Many search tools and algorithms have optimizations for fixed strings, i.e. static text.

Reviewer: What is the suggestion here, or future work, is it to consider these options? What even are the options to consider? Is it to somehow take advantage of SIMD instructions or something similar. In which case what if you use existing log shapes, kind of like a cache, where on hits you perform really well, and on missed you update the cache. The obvious problem being the matching part of the log is not constant.

However, Log Surgeon inherently works with patterns for non-static text;
in fact, it needs to identify matches among many different patterns of non-static text,
so we have not found meaningful opportunity for these optimizations.

> Currently, Log Surgeon does not report the earliest possible match for each rule;

Reviewer: Is the cost of finding these even worth it? What about some non greedy way to advanced the DFA, although I think we never found any success with this. Or some way to leverage the log shapes?

heuristically, it makes use of delimiter characters to achieve similar results.

There are a few known ways to implement earliest match semantics
(sometimes referred to as "leftmost-longest",
though "leftmost" also means something different in an overlapping context,
so we avoid that terminology).

Theoretically, the best option is a reverse scan for potential starting positions:

1. Construct a DFA for the reversed patterns, prefixed with `.*`.
2. In a single pass from the end to start of the file, mark the position at each accepting state.

Each of the marked positions necessarily corresponds to a valid start position for the "forwards" DFA.
So, starting from the first possible position, we simply run the forwards DFA to determine the rule match,
and the tagged DFA to determine the sub-rule matches as before.
For the next root match, we continue at the first starting position after our previous match ends.

The downsides of this approach are:

> - This approach necessarily requires reading from the end of the file.

Reviewer: ```suggestion
- Requires scanning from the end of the file.
```

- Reading in reverse is suboptimal compared to only reading forwards.
- Fully determinizing a DFA for a `.*`-prefixed regex is costly, though this can be worked around by lazy construction.
- Storing all starting positions is not necessarily prohibitive (e.g. using a bitmap), but not trivial either.

> ### Separating Log Events

Reviewer: Probably add a Log Even section and make this under that, and define what a log event is?

While lexing input, to determine log event boundaries, newline characters (not part of a rule match)
and root rules named `header` are treated specially under the following conditions.

A `header`, if preceded by a newline (or at start of input), is a log event separator.
Otherwise, it is treated as an ordinary rule.

If no separating `header` has been encountered (yet), log events are separated on newlines.

Newlines are part of the line preceding it;
in other words, log events are always terminated by newlines
(but a newline doesn't necessarily terminate a log event).

### TODO
Explain:
- anchors

> - leaf ambiguity

Reviewer: Resolve or move to git issues?
